### PR TITLE
Organ cell

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Refactors the function of the internal organ cell to be a "cell holder".

Cell holders, as the name implies, actually hold cells.

These deplete, over time (the default battery will take ~33 minutes to deplete), but can be recharged by using the APC plug-in power cord, or by utilizing a recharging station. If they're hit with an EMP, it will EMP the internal cell. Status of the cell is displayed 

IPCs have a special interaction with power holders; if they don't have one, it's severely damage, or the cell installed in it has run out of charge, the IPC will lose consciousness and begin taking damage.

This also adds in a `NO_HUNGER` species flag that does a number of small things: 

- Prevents hunger drain/gain
- Prevents benefitting from satiety/metabollic benefits

IPCs have `NO_HUNGER` by default, as they no longer use nutrition at all; they're dependent on power, from their internal cells.

Other changes:
- Recharge stations can now accept live/dead/unconscious mobs (should help with recharging IPCs who's batteries are drained without doing full surgery)

Code related
- Cleaned up the mess that is borg recharging code
- Arm recharger now properly uses cell relates procs instead of manipulating the `charge` variable

Potential thoughts/ideas:
- Being low on power slows you down?
- Stepping/moving drains power?

:cl: Fox McCloud
tweak: Internal cell organ replaced with a cell holder that holds an actual power cell.
tweak: IPCs will lose consciousness and take rapid damage if they don't have a cell or cell charge
tweak: IPCs no longer need nutrition
tweak: Cell alerts are given as your battery drains if you have a cell holder organ
/:cl:

### CODE REVIEW

**Comment by shazbot194** [comment]
I fully understand why having no charge would make an IPC become unconscious, but why would they take damage from it? Let alone “rapid damage”

**Comment by Jfrost6** [comment]
Why are we nerfing IPC's even more and shit doesn't take damage when their battery runs out. IPC's are gimped as it is, all they do is not breath air and are slightly easier to repair. But: they take extra damage, can't get genetic mutations, and now they explode if they don't charge for 30 minutes?

**Comment by Jfrost6** [comment]
If there's no reasonable way to make your changes work on the already working system in place for ipc's then why should we change to this, It's making a simple thing needlessly complicated.

**Comment by taukausanake** [comment]
> removing a vital organ is essentially meaningless; they'll die, then instantly revive

Isn't that part of their point? Easy to kill or "kill", easy to bring back online? IPCs only really die when their posibrain is destroyed. Anything else is basically just an inconvenience to them and an annoyance at best

**Comment by HereticChurch** [comment]
This damage that kills the IPCs when they run out of power, will it require extensive welderfuel/wires to repair, or is it just a matter of giving them a fresh cell or tossing them into a charge station? Not sure I'd look forward to the tedium of all the repairs because someone forgot to juice up at the local APC.

**Comment by Tayyyyyyy** [comment]
> For the same reason removing an IPC's internal cell kills them (which has been in place since forever); the implications is that no cell means no life; this is an extension of that, just that you have actual power now, instead of a pseudo-power via nutrition.

I get making things more lore/immersion friendly, but it seems like there could be better ways to do this that are less punishing gameplay wise. Generally it shouldn't be too difficult for an IPC to stay powered, but it does seem like this makes EMPs an even worse experience for IPCs. Their little power thingy already gets broken upon EMP, and now their internal cell will get damaged too, in addition to all their limbs falling off and their face becoming unrecognizable. Robotics can already be pretty questionable with repairing IPCs, so this could create very negative player experiences for little benefit.

**Comment by pinatacolada** [comment]
and if an ipc goes ssd for 30 minutes, it returns dead and exploded?

**Comment by shazbot194** [comment]
So someone brought up what about people who have to spend time away from APCs such as the pod pilot or mechanic? I think something appropriate to solve this would be to add the ability for IPCs to syphon energy off cells in this PR. Else we may end up with a lot of pilots being stranded inside their pod.

**Comment by Jfrost6** [comment]
@Allfd upsides of what? vulps? ipcs?

1.  if it's vulps well, They can see in the dark for 8 meteres or tiles which is a pretty huge buff, They have claws for sharp damage instead of blunt which i can't tell if its a bonus or what, and whats their debuff? they take extra damage from welding without eye protection, that's one of the weakest excuses for a nerf i've ever heard, the only people that could possibly effect are people who are newer or aren't paying attention in a low risk environment.

2.  The upsides of ipc's however are that they don't breathe and yet they still take pressure damage so they might be able to survive in space for a small amount of time over other races other than dionae. And they can also repair themselves on the go which I think is perfectly balanced considering they take a fuckton of extra damage from almost everything. And they still take brain damage from a huge source of toxin damage which is beer, they'll still stop working after too much just like anyone else except the effects are much worse. And then ipc's cant get genetic mutations making genetics nearly impossible for them and they have instakill items in the form of any emp. Labeling this pull with "Balance" is utterly hilarious to me as it's just making ipc's even worse to where they cant even be fun to play anymore, especially with the recent "balance" that makes ipc's take EVEN MORE damage. Ipc's aren't inherently good at anything, in a fight they'll be dead before they can repair themselves, minor damage is the only area where I can see ipc's advantages.

File: code/modules/mob/living/carbon/human/human.dm [hunk 1]
```diff
@@ -130,6 +130,7 @@
 		qdel(E)
 	for(var/obj/item/organ/O in internal_organs)
 		qdel(O)
+	death()
```
Comment (Crazylemon64) [inline code/modules/mob/living/carbon/human/human.dm]: Won't this cause a `seizes up lifeless` message when a human is deleted, like at round start for `robotize`

File: code/modules/surgery/organs/cell_holder.dm [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+			var/cellcharge = cell.charge/cell.maxcharge
+			switch(cellcharge)
+				if(0.75 to INFINITY)
+					owner.clear_alert("cell_organ")
+				if(0.5 to 0.75)
+					owner.throw_alert("cell_organ", /obj/screen/alert/lowcell, 1)
+				if(0.25 to 0.5)
+					owner.throw_alert("cell_organ", /obj/screen/alert/lowcell, 2)
+				if(0.01 to 0.25)
+					owner.throw_alert("cell_organ", /obj/screen/alert/lowcell, 3)
```
Comment (Crazylemon64) [inline code/modules/surgery/organs/cell_holder.dm]: Low charge should be better telegraphed - since it's lethal when it hits 0, with no other symptoms beforehand, I can foresee it sneaking up on players who're occupied with the rest of the round.

File: code/modules/surgery/organs/cell_holder.dm [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	..()
+
+/obj/item/organ/internal/cell_mount/Destroy()
+	QDEL_NULL(cell)
+	return ..()
+
+/obj/item/organ/internal/cell_mount/attackby(obj/item/I, mob/user, params)
+	if(isscrewdriver(I) && cell)
+		to_chat(user, "<span class='notice'>You remove [cell] from [src].</span>")
+		cell.forceMove(get_turf(src))
```
Comment (Crazylemon64) [inline code/modules/surgery/organs/cell_holder.dm]: Would a `put_in_hands` call if one of the hands is empty be a good idea?

File: code/modules/surgery/organs/cell_holder.dm [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		else
+			owner.throw_alert("cell_organ", /obj/screen/alert/nocell)
+
+/obj/item/organ/internal/cell_mount/emp_act(severity)
+	if(emp_proof)
+		return
+	if(cell)
+		cell.emp_act(severity)
+
+/obj/item/organ/internal/cyberimp/arm/power_cord
```
Comment (Crazylemon64) [inline code/modules/surgery/organs/cell_holder.dm]: With charge now being a life-and-death matter for IPCs, alternative forms of recharging that aren't dependent on station infrastructure may be a good idea - like an alcohol-electricity converter, or some kind of plasma or uranium engine/reactor (although those two might not be good options for roundstart, since those materials are rare and contested).

File: code/modules/mob/living/carbon/human/species/station.dm [hunk 5]
```diff
@@ -991,6 +991,16 @@
 		/mob/living/carbon/human/proc/change_monitor
 		)
 
+/datum/species/machine/handle_life(mob/living/carbon/human/H)
```
Comment (Crazylemon64) [inline code/modules/mob/living/carbon/human/species/station.dm]: I don't like how this is handled - it might be better to add an `on_life` for MMI holders that checks for the presence of a cell mount and its charge, and then call `death()` if there's no power available.
Then, you can add a proc, `is_autorevivable`, to brain organs - and override it on the MMI to carry over the code found [here](https://github.com/ParadiseSS13/Paradise/blob/master/code/modules/mob/living/carbon/human/human_damage.dm#L20) - as well as the check for a cell mount and power being available. That way, we can axe the `can_revive_by_healing` species flag entirely, and have it implemented entirely on organs.

File: code/modules/surgery/organs/cell_holder.dm [hunk 6]
```diff
@@ -0,0 +1,136 @@
+/obj/item/organ/internal/cell_mount
```
Comment (Crazylemon64) [inline code/modules/surgery/organs/cell_holder.dm]: Some proc to show additional information to the robot analyzer would be a good idea - that way, if a roboticist needs to find out what's gone wrong with a "Dead" IPC, they can see what's causing the IPC to not come back to life

File: code/modules/surgery/organs/cell_holder.dm (hunk 4 above)
Comment (shazbot194) [reply code/modules/surgery/organs/cell_holder.dm]: If we really want to go in depth, we could possibly make a few options like an ethanol engine, a cell syphon, or even a small engine like a Radioisotope Thermoelectric Generator (RTG) to either have as loadout options (either from a new option or the point system), printables, or give an item that lets them both pick and auto install. Maybe even give different systems benefits and downsides like the plasma heating up the air and releasing CO2, the ethanol engine letting out just a little CO2, and the RTG exiting radiation when on.

**Comment by Tayyyyyyy** [comment]
> So someone brought up what about people who have to spend time away from APCs such as the pod pilot or mechanic? I think something appropriate to solve this would be to add the ability for IPCs to syphon energy off cells in this PR. Else we may end up with a lot of pilots being stranded inside their pod.

I think this would be a flavorful way to further extend the cell concept. Also, I think IPC manufacturers would foresee potential issues with running out of power, and I think it would make sense for them to be stunned and go into "power saving mode", upon which they can slowly recharge. It doesn't make sense for a phone to become permanently damaged upon losing charge.

**Comment by alexkar598** [comment]
Do IPCs really need a nerf like this which is more anoying than anything else? It's bad game design to make complex and anoying stuff for no reason at all.

**Comment by Codydozer** [comment]
I appreciate the contributions you all make, but can we not take large amounts of time to specifically make a race even worse to play? IPC are already the hardest race to play and survive with. This seems like the classic: focus in the wrong area which happens to the best of us.

**Comment by Codydozer** [comment]
About the whole deat from power loss? This doesn't apply at all sorry. Unless you want to also make borgs get nuked from loss of power, this doesn't make sense as the bold "balance" tag states.

**Comment by Spacemanspark** [comment]
I don't like the idea of them taking damage when the entire cell is drained.
Everything else seems alright tho.

**Comment by Allfd** [comment]
> I think this would be a flavorful way to further extend the cell concept. Also, I think IPC manufacturers would foresee potential issues with running out of power, and I think it would make sense for them to be stunned and go into "power saving mode", upon which they can slowly recharge. It doesn't make sense for a phone to become permanently damaged upon losing charge.

It does though?

The switch to Lithium Ion from Nickle Metal Hydride came with a downside, while both batteries suffered from low charge damage, and oh no... reverse charging....   With Lithium Ions it got way way worse.

All lithium ion batteries (not all, some have external circuits) carry some form of charge control circuitry, they ensure that if a battery falls bellow its lower voltage, it can never be charged again effectively killing it.  Because if you charge one that fell below this voltage it can go all Samsung on you.

So fully charged is like, 4.2 Volts,  total discharge is 2.5,  the circuit never lets it fall below 2.5 (or often 3 volts if they were paranoid) from load.  However it can still self discharge, and if it dips much further the protection circuitry will put it into sleep mode.

Sleep mode is also known as "dead"  as the circuit won't let it wake up and take a charge and should be recycled. 

So in short, your phone won't become permanently damaged from loosing charge, its battery however will for sure.

**Comment by Codydozer** [comment]
@Allfd while that is true, how does this at all make IPC more enjoyable to play? Currently they take much more increased damage and can easily be instant-killed cheaply. It seems like just another reason along the long list of others telling me I shouldn't play a race that I find quite neat. Is rushing robotics every roundstart to get in line for a top tier power cell so they don't die in maintenance while repairing a wire truly the new meta Paradise has in its future?

**Comment by Allfd** [comment]
Does it run down so quickly that this is a concern? @Codydozer

**Comment by Crazylemon64** [comment]
@Allfd @Codydozer The numbers:
(Power drains at 1 unit every 2 seconds.)[https://github.com/Fox-McCloud/Paradise-1/blob/01edf4f6d232a11897554e6af6a658b8f22c9fba/code/modules/surgery/organs/cell_holder.dm#L41]
(The default cell has 1000 units of power.)[https://github.com/ParadiseSS13/Paradise/blob/f00d350596f8268154d893678fcdd64202ba4963/code/modules/power/cell.dm#L14]
Thus, you'll have 2 seconds/u * 1000u of lifespan without a recharge, or 2000 seconds - which is 1800 + 200 seconds = 30 * 60 + 3 * 60 + 20 seconds = 33 minutes and 20 seconds.

**Comment by Codydozer** [comment]
But why the damage at all in the first place... Just make them go into the hard crit state on the floor they can so easily enter without the damage. They are boned anyways once they fall to the floor in perma sleep. Why add the salt to the wound of now they get punished for being distracted? Why punish them at all? Is being passed out on the floor permanently until some random walks over to them on the off chance he/she don't think the IPC is just SSD not enough?

**Comment by Allfd** [comment]
By killing them, they can ghost or rejoin the round.  I'd they don't die, they can't ghost or rejoin.

**Comment by Crazylemon64** [comment]
Yeah, perma-KO instead of just making them "Dead" for the duration they lack power is harsher - they'd be left looking at a black screen for that period of time, that may or may not be indefinite.

**Comment by Codydozer** [comment]
I still prefer how it was with them just suffering the same slowness as every other race, save for borgs. If every race could die from starvation, that'd be one thing but this isn't that. I call for a separate PR adding starvation in full to all races equally rather than just to IPC.

**Comment by Tayyyyyyy** [comment]
I don't think species variety is bad. However, when this is primarily a lore based change, I think we can make the impact more flavor rather than as punishing as death. It doesn't make sense that an IPC's battery life is only 30 minutes anyways.

**Comment by Codydozer** [comment]
@Tayyyyyyy Species variety is good! But why target IPC specifically with a Starvation PR and.... not the rest of the races? It doesn't make sense for other races to be immune to starvation death while IPC are victim to it.

**Comment by Dinarzad** [comment]
I can't really agree with the full scope of the PR's changes.

The notion of IPCs have a normal power cell inside of them is fine by me, it even opens up the realm of 'upgrading' IPCs same as borgs, with newer, better power cells. Even Self-charging ones to be a sort of IPC equivalent to the 'No hunger' virus that fleshies can get.

Them DYING from no power though, I vehemently disagree with.  I can be down with IPCs running out of charge and possibly falling unconscious, or going blind and slow as they enter into emergency backup reserves until recharged. This puts an IPC in a bad way if they don't keep their charge up, worse then organics who aren't eating, but still are in a spot to resolve their problem.
But death is a bridge too far, particularly when it's a race that already has a 1-shot kill mechanic that can affect them through walls, on top of taking way more damage from the most common sources in the game, brute and burn. If the Cell charge = Death thing happens one of those two needs to be removed or drastically lowered to compensate.

**Comment by taukausanake** [comment]
IPCs are so squishy they might as well be the only race that can't die short of complete destruction. I think that's a nice trade off if the resistant to everything biological isn't already too much for them

**Comment by Codydozer** [comment]
@taukausanake arguably the immunity to all viruses is more of a burden because they cannot get the incredibly powerful healing viruses that can bring someone back up after licking Death's shoes.

**Comment by Allfd** [comment]
> arguably the immunity to all viruses is more of a burden because they cannot get the incredibly powerful healing viruses that can bring someone back up after licking Death's shoes.

Now that you bring it up, IPCs are very OP.

**Comment by shazbot194** [comment]
> Now that you bring it up, IPCs are very OP.

Have you ever had to heal each limb individually? Or do you often just apply one patch to your torso and call it good, or hop into a cryo cell, or be injected with something weird by the virologist and call it good for the entire round as nothing short of major trauma will actually do you in? 
Being immune to viruses is rather bad when 9 times out of 10 they actually heal you so much only rather robust people can actually kill you, and the other times, it makes you either sniffle, or have easily cure-able brain damage. Now compare that to a race that physically has to heal each limb individually, who has a one hit kill, and who takes an additional 65% damage and tell me how they are OP.

**Comment by Allfd** [comment]
> Or do you often just apply one patch to your torso and call it good, or hop into a cryo cell, or be injected with something weird by the virologist and call it good for the entire round as nothing short of major trauma will actually do you in?

Um, I am used to using patches on each limb, then waiting in line for surgery for half the round.  Get out of medbay, and find out that now all my internal organs are infected because Dr. GreyTide didn't wash their hands.

Or, shoved into cryo with massive internal bleeding, die, the defibrillator ran out of power, so off to cloning where the tube is out of bio-matter so they just shove you in the morgue.

None of this applies to Vox, who are just basically killed by medbay then gibbed with SR.


> Being immune to viruses is rather bad when 9 times out of 10 they actually heal you so much only rather robust people can actually kill you, and the other times, it makes you either sniffle, or have easily cure-able brain damage.

Healing viruses have had an uptick recently.  The majority of viruses I seem to encounter are bad news,  I can't recall being saved from death by a virus before.

> Now compare that to a race that physically has to heal each limb individually, 

most of them?

> who has a one hit kill, 

EMP again?  The only real significant IPC weakness, the one everyone seems to want to get rid of?

Oh wait, Vox, cortical stacks.  One hit EMP kill.

> and who takes an additional 65% damage 
Its 50% not 65%

Vox get 20% and you for sure can't knock their heads off and expect them to live.

> and tell me how they are OP.

They are OP.

**Comment by Chronarch2954** [comment]
> Oh wait, Vox, cortical stacks. One hit EMP kill.
Vox take at least 3 EMPs before they die, not one.

**Comment by Dinarzad** [comment]
@Allfd Really? That's the attitude we're gonna take with this?
Aight.

> Um, I am used to using patches on each limb, then waiting in line for surgery for half the round. Get out of medbay, and find out that now all my internal organs are infected because Dr. GreyTide didn't wash their hands.

You mean infection mechanics that aren't working? Curious.

> Or, shoved into cryo with massive internal bleeding, die, the defibrillator ran out of power, so off to cloning where the tube is out of bio-matter so they just shove you in the morgue.

So... you're saying people die or never get healed because they don't know how to properly treat you and just sorta shove you aside. Yup. IPCs sure do never experience that problem.

> None of this applies to Vox, who are just basically killed by medbay then gibbed with SR.

And who are also completely unrelated to anything in this topic at hand, so this is practically falling over yourself to reach for straws.

> Healing viruses have had an uptick recently. The majority of viruses I seem to encounter are bad news, I can't recall being saved from death by a virus before.

Then you don't play the same game as us.  It is **LITERALLY** Against the server rules for a non-antag virologist to dump out a death virus, and even an antag viro tends to need the hijack objective to do so without getting their ass in trouble. The main thing a Virologist can do REGULARLY is create some form of healing virus. If a Viro is in the round, it's almost as reliable to get a healing virus put out, as it is to get RnD maxed out.

> most of them?

That's weird. See, Stypitic powder/Silver Sulfa/Synthflesh does this thing where it actually uses a chemical, so it heals across the entire body, and the only time you do things limb by limb is if you're using Gauze/Trauma packs/Burn packs. Which is sorta weird to me, since you have a chemist on hand who makes the former by the bucket load in addition to cryotubes. So. Y'know.

> EMP again? The only real significant IPC weakness, the one everyone seems to want to get rid of?
Oh wait, Vox, cortical stacks. One hit EMP kill.

Yeah I mean if vox players can handle the instant 1 shot through walls death mechani-
Oh wait, vox take like 3 before it causes death. Welp.
If you;re going to add YET ANOTHER way to fall over dead that only applies to one race, that you, y'know, weigh that up with the literally 1-shot kill mechanic they already have, and put that into consideration.

> Its 50% not 65%
Vox get 20% and you for sure can't knock their heads off and expect them to live.

Our bad, you only die ludicrously faster, not hella ludicrously faster.
And maybe Vox aren't expected to live through that cuz they only take 20% more damage.

And actually, no one asked for IPCs to survive being headless either, that just kinda happened when they got turned into rock'em sock'em robots and all their vital bits were pushed into their chest. So y'know. You could just... punch their chest instead of their head and be just fine.

> They are OP.

Good talk, you behaved like a real life grown up today.
Now you wanna cut the unwarranted sass so we can have a legitimate conversation now, or we gonna keep the tone of this whole thing at a solid "Angry teenager" ?

**Comment by shazbot194** [comment]
Thanks @Dinarzad, but I’m going to disprove this 50% myth. Yes, a comment does say 150%, but they clearly can’t do math. [if you look over at this PR](
https://github.com/ParadiseSS13/Paradise/pull/8894) you may notice [this line](
https://github.com/ParadiseSS13/Paradise/pull/8894#diff-c263363df08edeb8a6e85abb9a411321R934). Yes, this PR buffs IPCs to 150% total mod. Now, how is this possible, let me take you through some math. 2.28<2.5, so this must mean the original mod is higher then 150%, Just how much, we can do the math, .66 x 2.5 != 1.5, but rather == 1.65 == 165%, so do you want to keep telling me, someone who is more then familiar with this code I don’t know what I am talking about, and if you want to double check, shoot an IPC with a .357. It should do 90 damage, but rather does 99.

**Comment by Codydozer** [comment]
I just heard IPC were OP. I had a good smile today and a bit of a giggle strike me. 🤣

**Comment by Crazylemon64** [comment]
OK. I'm right here.
Anyways, *PLEASE* tone down the discussion, or we shall take moderative action.

**Comment by Codydozer** [comment]
@Crazylemon64 I think instead of starvation mechanics for IPC and **only** IPC are ridiculous when no other races have it. Instead, make their limbs not account to their total health and make them take that actual 150% brute and burn damage. We do not want more nerfs! 👎 👎

**Comment by Crazylemon64** [comment]
Can't say I didn't warn you.
